### PR TITLE
Fix and Enhance Usergroup ACL Permissions Grids and Filtering

### LINKED ISSUE #16239: Access Policy Context not selectable
### Summary
On an almost fresh MODX installation, the Context policy is not selectable on usergroup/update -> contexts. 

The connector of that combo uses the following properties:

start: 0
limit: 20
action: Security/Access/Policy/GetList
group: Administrator,Object
combo: true
query:

### Environment
MODX 3.0.1, PHP 7.4

### PR DESCRIPTION
**Pull request by smg6511**
### What does it do?
Made a number of changes and additions in the processors and grid JS files to correct issues with the grids’ grouping functionality and more accurately build the grid filters depending on the data shown. More specifically:

1. Replaced the column model renderers (used to render links on column data) with a `templatecolumn` specification (generated by a new method in the grids base class), as using renderers on grouping grids interferes with the grouping functionality (because a renderer is an interceptor method in Ext).
2. Moved menu specifications (grid context menus) out of processors and into the grid js for all sections (context was already this way). The primary reason being that some events (such as `afterRemoveRow`) were not firing when set up this way; monitoring these events is necessary for maintaining accurate filter options when making changes to a grid's data. Also, there's no reason to specify js in php unless the item in question will change on the fly and needs setup data from the processor.
3. Grid view, including the bottom toolbar, now correctly updates when a row is deleted.
4. Created new methods to update dependent filters (when two or more filters are present) based upon the selected value of any given filter.
5. Fixed some filters that were missing options due to an incorrect specification of their `baseParams.group` property.

### Why is it needed?
Grouping in these grids was not working correctly. Also, providing filter options where no relevant data is available in a given grid is a UX problem that needs to be fixed not only here, but across all grids with filters. 

### How to test
1. Ensure modx and browser caches are fully cleared.
2. Create a number of policies in each area (if you don't already have some set up) and play with the grouping and filtering of items in each grid. (These are the five Access Permissions grids — Context, Resource Groups, Element Categories, Media Sources, and Namespaces.) Verify that filters only show relevant options depending on the grid data and other filters you've made selections on.
3. Verify that other operations (editing and deleting) affect filter options as expected.

### Related issue(s)/PR(s)
Resolves #16239.

### Special Note
Because these updates involve a lot of changes, I didn't want to execute them for _all_ relevant grids at once. This PR completes a logical section and, depending on any requested changes, other PRs will follow what I've done here for other grids.

### Before and After Examples
Currently, in MODx 3.0.3-dev, note each of the following problems (in order of appearance in the video):

1. Each ACL area grid does not group properly; where an area's rows should be grouped under a common heading, they are not. To see this clearly, pause each time I click on each area and note where grouping should be occurring (for example, in Contexts, the three assigned the _Editors - 1000_ Role should be grouped under one row heading for that Role).
2. [0:19] Next, notice how the grid filters show irrelevant options where they should only show those present in the grid data. The _Filter by Context_ combo shows all Contexts instead of those present (dupctx1, dupctx2, dev2, and www2). The same goes for the _Filter by Policy_ combo. [0:42] Additionally, the Filter by Policy combo is mis-configured in places; here the problem in the Contexts area is illustrated. When adding a new Context, you are given the full Policies list in the Create window, but the filter combo shows an incomplete subset. [1:02] Lastly, when attempting to create additional policies for the same Context, an error occurs due to an incorrect validation for a unique context key. 
3. [1:32] When deleting a row, see how the count is not updated in the lower right grid status area.
4. [1:50] Next, when filtering, a choice in one combo should update the others to show only entries relevant to the current state of the query. They do not currently react to changes as one would expect.

https://user-images.githubusercontent.com/689075/209720912-944ca7bb-2d91-4aae-9bdd-f0b5b7d71b18.mov

Now, note how each problem is fixed (action taken in same order as above):

https://user-images.githubusercontent.com/689075/209720965-f3b53940-7554-4db1-a695-86bc2b25c8b7.mov

### CODE REVIEW

File: core/src/Revolution/Processors/Context/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Context/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

I think it's better to make this variable protected instead of private, especially since people can extend this class and the variable is used in public methods. But if you have a good argument to make them private, let me know!

File: core/src/Revolution/Processors/Context/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Context/GetList.php:108]: Have you considered to use an early return of `$c` instead of using a huge if statement block?

File: core/src/Revolution/Processors/Element/Category/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Element/Category/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

File: core/src/Revolution/Processors/Security/Access/Policy/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Security/Access/Policy/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

File: core/src/Revolution/Processors/Security/Access/Policy/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Security/Access/Policy/GetList.php:165]: Have you considered to use an early return of `$c` instead of using a huge if statement block?

File: core/src/Revolution/Processors/Security/ResourceGroup/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Security/ResourceGroup/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

File: core/src/Revolution/Processors/Security/ResourceGroup/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Security/ResourceGroup/GetList.php:73]: Have you considered to use an early return of `$c` instead of using a huge if statement block?

File: core/src/Revolution/Processors/Source/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Source/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

File: core/src/Revolution/Processors/Workspace/PackageNamespace/GetList.php
Comment (JoshuaLuckers) [inline core/src/Revolution/Processors/Workspace/PackageNamespace/GetList.php]:
```suggestion
    protected $isGridFilter = false;
```

**Review by JoshuaLuckers — COMMENTED** [review]
This is a huge PR with only 1 commit making it pretty hard for me to review at once. 
I've only reviewed a part of the code for now (without testing the functionality).

File: core/src/Revolution/Processors/Context/GetList.php
Comment (smg6511) [reply core/src/Revolution/Processors/Context/GetList.php]: Good point, I wasn't considering the possible extension of these GetList classes. I'll change that throughout. 
UPDATE: Change done for all 6 files.

File: core/src/Revolution/Processors/Context/GetList.php
Comment (smg6511) [reply core/src/Revolution/Processors/Context/GetList.php:108]: There have been other PRs I've worked on where an early return of X was suggested, helping me to learn the benefit of doing so in certain situations. I don't know that it really helps here. To me, doing what I think you're alluding to ...
```
if (!$this->isGridFilter) {
    return $c;
}
if ($userGroup = ...) {
[...]
return $c;
}

```
adds another couple lines and puts that return in a place you might not at first spot when scanning the code. Since an early return doesn't short circuit the parsing of any code, I don't see the benefit. That said, if you feel strongly about it (or obviously if I'm just missing something) I certainly don't mind changing it.

**Comment by smg6511** [comment]
@JoshuaLuckers - Thanks for beginning to take a look at this! I know it's relatively big, but in the past I've gotten caught between "it's covering too much" and "why did you only apply it to this one little area" (paraphrasing, and not referring to any particular reviewer). I was hoping this would be a happy medium ;-)

I think once you get a closer look you'll see the repeating pattern of changes for the 5 grids, making it a little easier to assess. As with most of my bigger PRs, it seeks to centralize shared routines in the base classes to try to get away from all the needless duplication of similar and often times identical code across related classes.

Also note that I've resisted including other fixes I've identified for these grids, and will have a couple forthcoming PRs that make other incremental improvements that build on the ones here.

File: core/src/Revolution/Processors/Context/GetList.php
Comment (JoshuaLuckers) [reply core/src/Revolution/Processors/Context/GetList.php:108]: I don’t have a strong opinion on this, I was just wondering if you did consider the early return and you did.

**Comment by JoshuaLuckers** [comment]
> @JoshuaLuckers - I know it's relatively big, but in the past I've gotten caught between "it's covering too much" and "why did you only apply it to this one little area" (paraphrasing, and not referring to any particular reviewer). I was hoping this would be a happy medium ;-)

It sure is a happy medium, but in this case it would be helpful if the PR would consist of multiple commits. Especially since you described 5 different things this PR does.

The amount of changes we have to review would stay the same but reviewing small changes per commit is much easier (at least for me). We could also ignore certain ch ages easier (like code style changes). It also gives more context to certain code changes and it would make it easier to revert only a portion of the changes when needed instead of everything.

> Also note that I've resisted including other fixes I've identified for these grids

🙈❤️


For me it's fine to leave it as is and so we can go back on topic ;)

File: core/src/Revolution/Processors/Context/GetList.php
Comment (mishanthrop) [inline core/src/Revolution/Processors/Context/GetList.php:98]: For what cases is explicit type casting here?

Comment (smg6511) [reply core/src/Revolution/Processors/Context/GetList.php:98]: Although the usergroup property _should_ only be exposed internally, casting it to int ensures only a safe value will be sent in the query. Note that this format of the join's conditions (a non-associative array of strings) does not get sanitized.

**Comment by JoshuaLuckers** [comment]
I tried to understand how to test it and what I should see.

Would it be possible to make a small video demonstrating it?

**Comment by smg6511** [comment]
@JoshuaLuckers - Yes, I'll put a couple before and after videos to demonstrate the changes. Give me a day or so...

**Comment by JoshuaLuckers** [comment]
> @JoshuaLuckers - Yes, I'll put a couple before and after videos to demonstrate the changes. Give me a day or so...

No hurries!

**Comment by smg6511** [comment]
@JoshuaLuckers - Ok, I put a couple videos up in the main PR description.

**Comment by JoshuaLuckers** [comment]
> @JoshuaLuckers - Ok, I put a couple videos up in the main PR description.

Thanks a lot, now I know what I should look for when testing haha.

**Comment by smg6511** [comment]
One thing to be aware of: A while after submitting this PR I discovered that performing additional sorting on the grids made the grouping fall apart. I do have that figured out and will submit that fix in a new PR once this one is incorporated.

**Review by JoshuaLuckers — APPROVED** [review]

**Comment by smg6511** [comment]
@Ibochkarev - I see you've been doing some reviews lately and was hoping you might have some time to take a look at this. I've got some follow-up work that depends on this PR, so would love to see it merged sooner than later.

**Review by Mark-H — APPROVED** [review]
Giving this the 👍 based on your very detailed test and a detailed look through the code. You've clearly spent a lot of time and effort on this and it fixes a number of issues, so well done and thank you. 

The only comment I have is that it would be good to prepare some [documentation](https://github.com/modxorg/docs) for the new utility methods on `MODx.Grid`. Those can be useful for others, and we'll also need to have that information as core team in the future when we forgot this PR and need to figure out how it worked ;) 

There may be minor issues or edge cases I've overlooked, but we can address those in the future.

**Comment by smg6511** [comment]
@Mark-H - Ok, I'll put documentation on my near-term agenda. Thanks for reviewing!

**Comment by opengeek** [comment]
Ok, so this has a ton of conflicts after merging #16089. Can these be resolved today so we can include this in 3.0.3 tomorrow?

**Comment by smg6511** [comment]
@opengeek - Yes, I'll take a look at those in just a bit...

**Comment by smg6511** [comment]
@opengeek - Ok, now everything's working together nicely.

**Comment by opengeek** [comment]
For future reference, you need to rebase your commits on the latest, not merge the new ones into them.

**Comment by opengeek** [comment]
This is going to have to wait for the next release, as there is no way to sort through this mess like this. Can you please revert the merges and rebase the work from just this PR on the up-to-date 3.x branch? When you have time…

**Comment by smg6511** [comment]
@opengeek - I know about the preference to rebase and actually _tried_ to do just that but ran into issues. The changes and conflicts were so intertwined that it seemed to be too confusing for rebase to deal with. It was weird, I started off in the rebase process and ended up to my surprise with a merge happening. (Note that I use Github Desktop and VS Code.) Anyway, given that background, is it easier if I just pull this apart and resubmit in a new PR?

**Comment by Mark-H** [comment]
Don't mind me, [just dropping a useful link for anyone looking to get better at rebases](https://www.youtube.com/watch?v=uI1V7771plw).

**Comment by smg6511** [comment]
@Mark-H - Thanks, I _will_ check that out. I usually don't run into issues and have rebased many of my PRs and in-progress work, but this one was a different story!

**Comment by Mark-H** [comment]
Closing in favour of #16355

**Pull request closed without merging**
